### PR TITLE
Namespace analyzer resolves #30626

### LINKED ISSUE #30626: Need analyzer for conflicting namespace level resources
For example:
```yaml
apiVersion: v1
items:
- apiVersion: security.istio.io/v1beta1
  kind: PeerAuthentication
  metadata:
    name: default
    namespace: istio-system
  spec:
    mtls:
      mode: STRICT
- apiVersion: security.istio.io/v1beta1
  kind: PeerAuthentication
  metadata:
    name: default2
    namespace: istio-system
  spec:
    mtls:
      mode: DISABLE
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
```

One will be ignored. This applies to many types, not just PeerAuthentication - we have one just for sidecar in https://preliminary.istio.io/latest/docs/reference/config/analysis/ist0111/

### PR DESCRIPTION
**Pull request by dbgoytia**
[ X] User Experience

Pull Request Attributes

Please check any characteristics that apply to this pull request.

[ ] Does not have any user-facing changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

Subject: Need analyzer for conflicting namespace level resources (#30626) to check namespace level resources (PeerAuthentication, AuthorizationPolicy, RequestAuthentication)
Description:

    Adds the capability to review namespace level conflicts for these resources.
    Checks for conflicts with selectors of these resources.
    Checks for conflicts with the selector of this resource with namespace wide configurations.

Todo:
* Consolidation of this analyzer with sidecar.SelectorAnalyzer since they are performing alike tasks.
* Add a mesh level analyzer (which I'm not yet already sure it's implemented)
* Consider adding revisions for this resources on K8s.

### CODE REVIEW

**Comment by dbgoytia** [comment]
Reference back to this conversation for full details and past comments:
https://github.com/istio/istio/pull/32929

**Comment by dbgoytia** [comment]
@shamsher31 @howardjohn, this is the new version of the PR I placed together previously, and fixed the styling issues I was having previously.

File: galley/pkg/config/analysis/analyzers/namespaceconflict/namespaceselector.go
Comment (howardjohn) [inline galley/pkg/config/analysis/analyzers/namespaceconflict/namespaceselector.go:68]: I think @yangminzhu mentioned AuthorizationPolicy actually are merged so this may not be applicable for this one. But the other two are

File: galley/pkg/config/analysis/analyzers/namespaceconflict/namespaceselector.go
Comment (howardjohn) [inline galley/pkg/config/analysis/analyzers/namespaceconflict/namespaceselector.go:54]: should we consolidate with sidecar.SelectorAnalyzer? I think its doing the same thing

Comment (howardjohn) [reply galley/pkg/config/analysis/analyzers/namespaceconflict/namespaceselector.go:54]: We can do it in a followup though

Comment (howardjohn) [reply galley/pkg/config/analysis/analyzers/namespaceconflict/namespaceselector.go:54]: That one also seems to have a nice property that it checks that at least one pod is selected, so you know your policy has no effect

File: galley/pkg/config/analysis/analyzers/testdata/namespaceconflict-reqauth-conflicts.yaml
Comment (howardjohn) [inline galley/pkg/config/analysis/analyzers/testdata/namespaceconflict-reqauth-conflicts.yaml:71]: I don't think this should, the precedence is root namespace > namespace level > workload selector.
So its fine to have a namespace level one and a workload selector one, the latter will take precedence

File: galley/pkg/config/analysis/analyzers/namespaceconflict/namespaceselector.go
Comment (dbgoytia) [reply galley/pkg/config/analysis/analyzers/namespaceconflict/namespaceselector.go]: Hey @howardjohn @yangminzhu , awesome! I think it makes perfect sense, maybe we can do different type of analysis for that particular resource, I've removed it from this PR.

File: galley/pkg/config/analysis/analyzers/namespaceconflict/namespaceselector.go
Comment (dbgoytia) [reply galley/pkg/config/analysis/analyzers/namespaceconflict/namespaceselector.go:53]: I think that's a good idea, I haven't seen the code over in sidecar.SelectorAnalyzer, it seems to have a logic that's very alike to this analyzer, how would you prefer to move forward? 🤔

File: galley/pkg/config/analysis/analyzers/testdata/namespaceconflict-reqauth-conflicts.yaml
Comment (dbgoytia) [reply galley/pkg/config/analysis/analyzers/testdata/namespaceconflict-reqauth-conflicts.yaml]: Thanks a lot for noticing that John, I think I fixed the issue, right now it's reporting the issues as follows:



For workloads that don't have a selector associated (namespace wide) (peer authentications as an example)
```
Warning [IST0146] (PeerAuthentication defaultconfig.default) More than one istio/security/v1beta1/peerauthentications in namespace "default" for workload "(ALL) Namespace: default": [default/defaultconfig default/defaultconfig2].
Warning [IST0146] (PeerAuthentication defaultconfig2.default) More than one istio/security/v1beta1/peerauthentications in namespace "default" for workload "(ALL) Namespace: default": [default/defaultconfig default/defaultconfig2].
```


And for workloads that have a selector associated (selector conflicts) (request authentications as an example)

```
Warning [IST0146] (RequestAuthentication seldummybar.bar) More than one istio/security/v1beta1/requestauthentications in namespace "bar" for workload "app=dummy": [bar/seldummybar bar/seldummybar2].
Warning [IST0146] (RequestAuthentication seldummybar2.bar) More than one istio/security/v1beta1/requestauthentications in namespace "bar" for workload "app=dummy": [bar/seldummybar bar/seldummybar2].
```

I removed that test scenario and the test cases I defined are working as expected. Please do let me know if you think I should add more validation or some other cases that I might be missing! :) 🤔

File: galley/pkg/config/analysis/analyzers/namespaceconflict/namespaceselector.go
Comment (dbgoytia) [reply galley/pkg/config/analysis/analyzers/namespaceconflict/namespaceselector.go:53]: I added a bit of logic so we can easily consolidate them together using this :) :
```
c.ForEach(<TYPE>, func()  {
  if x.GetSelector() != nil {
	  // If the resource has workloads associated with it, analyze for conflicts with selector
	  a.analyzeWorkloadSelectorConflicts(r, c)
  } else {
	  // If the resource has workloads associated with it, analyze for conflicts with namespace
	  a.analyzeNamespaceWideConflicts(r, c)
  }
})
```

Comment (dbgoytia) [reply galley/pkg/config/analysis/analyzers/namespaceconflict/namespaceselector.go:53]: BTW, I updated the to-do list to consolidate with this other analyzer :) @howardjohn !

File: galley/pkg/config/analysis/analyzers/namespaceconflict/namespaceselector.go
Comment (shamsher31) [inline galley/pkg/config/analysis/analyzers/namespaceconflict/namespaceselector.go:38]:
```suggestion
	peerAuthCollection    = collections.IstioSecurityV1Beta1Peerauthentications.Name()
```

Comment (shamsher31) [reply galley/pkg/config/analysis/analyzers/namespaceconflict/namespaceselector.go:38]: Just a convention that we are using across analyzers

File: galley/pkg/config/analysis/analyzers/namespaceconflict/namespaceselector.go
Comment (shamsher31) [inline galley/pkg/config/analysis/analyzers/namespaceconflict/namespaceselector.go:39]:
```suggestion
	requestAuthCollection = collections.IstioSecurityV1Beta1Requestauthentications.Name()
```

File: galley/pkg/config/analysis/analyzers/namespaceconflict/namespaceselector.go
Comment (shamsher31) [inline galley/pkg/config/analysis/analyzers/namespaceconflict/namespaceselector.go:44]:
```suggestion
		Name:        "namespaceConflict.Analyzer",
```

File: galley/pkg/config/analysis/analyzers/namespaceconflict/namespaceselector.go
Comment (shamsher31) [inline galley/pkg/config/analysis/analyzers/namespaceconflict/namespaceselector.go:108]: This message can be better.

File: galley/pkg/config/analysis/analyzers/namespaceconflict/namespaceselector.go
Comment (shamsher31) [inline galley/pkg/config/analysis/analyzers/namespaceconflict/namespaceselector.go:147]: Better error default please.

File: releasenotes/notes/30626.yaml
Comment (shamsher31) [inline releasenotes/notes/30626.yaml:8]:
```suggestion
  **Added** a namespace level analyzer for `PeerAuthentication`, `RequestAuthentication` and `AuthorizationPolicy`.
```

File: galley/pkg/config/analysis/analyzers/namespaceconflict/namespaceselector.go
Comment (dbgoytia) [reply galley/pkg/config/analysis/analyzers/namespaceconflict/namespaceselector.go:108]: Hi @shamsher31 I added a different default message, but still I'd be happy to get your feedback or in case you have any suggestion it would be great!

**Comment by dbgoytia** [comment]
Hi @shamsher31 ! I fixed the styling issues that you kindly suggested

**Comment by dbgoytia** [comment]
@googlebot I fixed it

**Comment by dbgoytia** [comment]
@googlebot I fixed it

**Comment by dbgoytia** [comment]
@googlebot I fixed it

**Comment by dbgoytia** [comment]
Hi @shamsher31 :D. I think I refactored wrongly the package name from namespaceconflict -> namespaceConflict, so there were some errors in CI, after that, I did the refactoring again just to be sure everything was ok, and got into this message from the linter:

```
INFO [runner] linters took 1m57.5472263s with stages: goanalysis_metalinter: 1m55.3026724s
galley/pkg/config/analysis/analyzers/namespaceConflict/namespaceselector.go:15:1: don't use MixedCaps in package name; namespaceConflict should be namespaceconflict (golint)
package namespaceConflict
```

So I re-changed that to "namespaceconflict" for now to be able test. So far I think they are passing, just missing the "Tide" test.

**Review by therealmitchconnors — APPROVED** [review]

**Comment by howardjohn** [comment]
I think the release note issue occurs when a note was added them removed. Not 100% sure. You might just need to rebase and it will work again?

**Comment by dbgoytia** [comment]
Hey @howardjohn not sure if this was the right way to go, but found it and added it to a new PR using git checkout

https://github.com/istio/istio/pull/34260

Or is there a better way?

**Comment by ericvn** [comment]
Normally a rebase on the current head will pull the intermediate commits of the chain where the file was added and then deleted. As I noted in the other PR, there is work to fix the pipeline so that it only finds changes in this PR and not the others.

**Comment by dbgoytia** [comment]
I think I kind of messed up :(

I did a:

```
git fetch
git checkout namespace-analyzer-30626
git rebase origin/master
```

And now I have a ton of commits in my branch that were supposed to be already merged

**Comment by brian-avery** [comment]
Happens to the best of us :) . If you do a `git reflog`, you should be able to find your commit from before you did the rebase and check that out.

**Comment by dbgoytia** [comment]
Im not sure why the **integ-istiodremote-k8s-tests_istio** is flaking, sometimes when I re-run the test it passes 🤔 @howardjohn do you have any ideas?

**Comment by ericvn** [comment]
@dbgoytia Let me look into the test flakes. We can make it optional if needed.

**Comment by ericvn** [comment]
The istiodremote test failure sees to be related to your conflicting files. It is now optional.
```
$ git merge --no-ff fc66814a5abf21c99557200ebada4838c9eea0c8
Auto-merging galley/pkg/config/analysis/msg/messages.yaml
CONFLICT (content): Merge conflict in galley/pkg/config/analysis/msg/messages.yaml
Auto-merging galley/pkg/config/analysis/msg/messages.gen.go
CONFLICT (content): Merge conflict in galley/pkg/config/analysis/msg/messages.gen.go
Auto-merging galley/pkg/config/analysis/analyzers/analyzers_test.go
CONFLICT (content): Merge conflict in galley/pkg/config/analysis/analyzers/analyzers_test.go
Automatic merge failed; fix conflicts and then commit the result.
# Error: exit status 1
```

**Comment by dbgoytia** [comment]
awesome!! thanks a lot, @ericvn, really appreciate your help. I rebased my commit

File: galley/pkg/config/analysis/analyzers/namespaceconflict/namespaceselector.go
Comment (shamsher31) [inline galley/pkg/config/analysis/analyzers/namespaceconflict/namespaceselector.go:93]:
```suggestion
```

File: galley/pkg/config/analysis/analyzers/namespaceconflict/namespaceselector.go
Comment (shamsher31) [inline galley/pkg/config/analysis/analyzers/namespaceconflict/namespaceselector.go:101]:
```suggestion
```

File: galley/pkg/config/analysis/analyzers/namespaceconflict/namespaceselector.go
Comment (shamsher31) [inline galley/pkg/config/analysis/analyzers/namespaceconflict/namespaceselector.go:108]:
```suggestion
			fmt.Printf("Unsupported type:  %T!\n", v)
```

File: galley/pkg/config/analysis/analyzers/namespaceconflict/namespaceselector.go
Comment (shamsher31) [inline galley/pkg/config/analysis/analyzers/namespaceconflict/namespaceselector.go:147]:
```suggestion
		fmt.Printf("Unsupported type:  %T!\n", v)
```

File: galley/pkg/config/analysis/analyzers/namespaceconflict/namespaceselector.go
Comment (shamsher31) [inline galley/pkg/config/analysis/analyzers/namespaceconflict/namespaceselector.go:233]:
```suggestion
		fmt.Printf("Unsupported type:  %T!\n", v)
```

File: releasenotes/notes/30626.yaml
Comment (shamsher31) [inline releasenotes/notes/30626.yaml:8]:
```suggestion
  **Added** a namespace level analyzer for `PeerAuthentication` and `RequestAuthentication`.

```

Comment (dbgoytia) [reply releasenotes/notes/30626.yaml]: Thanks a lot @shamsher31 for taking the time to review my PR. I added the changes you suggested  👍🏽  let me know if you think there's something else I should refactor.

**Review by shamsher31 — COMMENTED** [review]
LGTM, please fix the tests.

**Comment by dbgoytia** [comment]
Awesome! @shamsher31  just out of curiosity,  I can see the tide "pending - not mergeable" now, and if I click there I can see that there are other PR's queued for merging. After I complete whatever changes are still required, how do I schedule this PR for merging?

**Comment by ericvn** [comment]
@dbgoytia Once you get to tide-Pending and no Required tests not passing, all that's needed is the approval. In this case you'll see it's `Waiting on code owner review from istio/wg-networking-maintainers.`.  Once a networking person approves, it should merge (assuming something like a needs-rebase doesn't happen).

**Comment by dbgoytia** [comment]
Awesome! Thanks a lot @ericvn , I think I'll reach out in slack and see if someone from those groups can help take a look

**Review by shamsher31 — APPROVED** [review]

**Review by elfinhe — APPROVED** [review]

File: galley/pkg/config/analysis/analyzers/namespaceconflict/namespaceselector.go
Comment (yangminzhu) [inline galley/pkg/config/analysis/analyzers/namespaceconflict/namespaceselector.go:68]: Actually multiple request authn for same workload is allowed, we just merge them so that all request authn policies will have effect at the same time.

Unlike the peer authN, there is no conflict here as it is just to accept multiple issuers configured at the same time.

**Pull request closed without merging**
